File: projects/engineering.purdue.edu/kak/scriptingwithobjects/swocode/chap3/ForkWait.py
#!/usr/bin/env python

### ForkWait.py

import os                       # for fork, wait                    #(A)
import time                     # for strftime, sleep               #(B)
import commands                 # for getoutput                     #(C)


def print_processes( arg ):                                         #(D)
    child_or_parent = arg                                           #(E)
    now = time.strftime('%X')                                       #(F)
    print "\n"                                                      #(G)
    process_list = commands.getoutput( 'ps' ).split('\n')           #(H)
    for line in process_list:                                       #(I)
        print child_or_parent, str(now), str(os.getpid()), line     #(J)

child_pid = os.fork()                                               #(K)

if (child_pid == 0):                                                #(L)
    # Execute in child process only:
    print_processes( 'child:' )                                     #(M)
else:                                                               #(N)
    # Execute in parent process only:
    time.sleep(2)                                                   #(O)
    print_processes( 'parent:' )                                    #(P)
    zid = os.wait()                                                 #(Q)
    print_processes( 'parent:' )                                    #(R)
    print "\nReturned by wait: %s\n" % str(zid)                     #(S)

File: projects/demos/engineering.purdue.edu/scriptingwithobjects/swocode/chap2/ForkWait.pl
#!/usr/bin/perl -w

### ForkWait.pl

use strict;

my $child_pid = fork();                                            #(A)

die "Unable to fork: $!" unless defined $child_pid;                #(B)

if ($child_pid == 0) {                                             #(C)
    # Execute in child process only:
    print_processes( 'child ' );                                   #(D)
} else {                                                           #(E)
    # Execute in parent process only:
    sleep(2);                                                      #(F)
    print_processes( 'parent' );                                   #(G)
    my $zid = wait();                                              #(H)
    print_processes( 'parent' );                                   #(I)
    print "Returned by wait: $zid\n";                              #(J)
}

sub print_processes {                                              #(K)
    my ($child_or_parent) = @_;                                    #(L)
    my ($now) = (localtime =~ /(..:..:..)/);                       #(M)
    print "\n";                                           
    print map { "[$child_or_parent] [$now] [$$] $_" } `ps`;        #(N)
    print "\n"; 
}
